h2o-genmodel/src/main/java/hex/genmodel/algos/coxph/CoxPHMojoModel.java

```diff
@@ -0,0 +1,60 @@
+package hex.genmodel.algos.coxph;
+
+import hex.genmodel.MojoModel;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+public class CoxPHMojoModel extends MojoModel  {
+  double[] _coef;
+  int _numStart;
+  Map<List<Integer>, Integer> _strata;
+  int _strata_len;
+  double[][] _x_mean_cat;
+  double[][] _x_mean_num;
+  int[] _coef_indexes;
+
+  CoxPHMojoModel(String[] columns, String[][] domains, String responseColumn) {
+    super(columns, domains, responseColumn);
+
+  }
+
+  @Override
+  public double[] score0(double[] row, double[] predictions) {
+    double result = 0.0;
+
+    final int size = 0 < _strata.size() ? _strata.size() : 1;
+    double[] lpBase = new double[size];
+    for (int s = 0; s < size; s++) {
+      for (int i = 0; i < _x_mean_cat[s].length; i++)
+        lpBase[s] += _x_mean_cat[s][i] * _coef[i];
+      for (int i = 0; i < _x_mean_num[s].length; i++)
+        lpBase[s] += _x_mean_num[s][i] * _coef[i + _numStart];
+    }
+    
+    for (int i = 0; i < _coef_indexes.length; i++) {
+      final int coefIndex = _coef_indexes[i];
+      result += row[coefIndex] * _coef[i];
+    }
+    
+    result -= lpBase[strataForRow(row)];
+    
+    predictions[0] = result;
+    return predictions;
+  }
+  
+  private int strataForRow(double[] row) {
```

Comment:
we would ideally like to avoid memory allocation when scoring each on this API level, any idea how to avoid this?

Reply:
It depends on how score0 can be called. If I am promised that it happens from a single thread sequentially, it's easy, just move the arrraylist as a prealocated instance field.

In theory I can make an easy to understand data structure to replace `Map<List<Integer>, Integer> _strata;` but I am not sure it's worth it. I would replace an GC friendly allocation + deallocation (there is only one local refference to it with no escape) with some more memory jumping.

Reply:
@michalkurka Your opinion?

Reply:
there is no guarantee score0 is used in single-threaded environment, I think our MOJOs are supposed to be thread-safe

Reply:
I dont think a custom data-structure is necessary, maybe just a custom key object for the hashmap that could be initialized with an "oversized" array but still return correct hashCode and equals result just for the strata sub-array

Reply:
There is a way to do it relatively easily by implementing a simple hashtable, the Stratas are known in advance and there is a (relatively) a small number of them. You can calculate a hashcode for each strata identifier (composite key), sort them and do a binary-search for an incoming row, then you just compare the identifier elements. This can be encapsulated in a tiny hashtable class.

Your code is also auto(un)boxing the integers, however, this will be optimized.

Just an idea.